Simplify and optimize __fill_random_buf

gsauthof opened the pull request:
This reduces the number of source lines and the code size.

For example, when compiling with GCC 12.1 (-O3 -march=skylake), the
resulting assembly shrinks from 33 to 27 instructions and the number of
jump instructions is reduced from 4 to 3.

NB: GCC is able to eliminate the memcpy() call.

NB: Even if a compiler doesn't eliminate the memcpy() call, it's very
unlikely to ever get called since the buffer sizes are expected to be
powers of two [greater than 8], usually.

---

Motivation: On a somewhat underpowered CPU I noticed that fio was CPU bound (when running with `refill_buffers`) and profiling showed that most of the time was spent in `__fill_random_buf()`.

I thus optimized it a bit and observed an improvement of 20 % or so in the fio throughput.

See also: https://godbolt.org/z/57fGGYqsM

axboe commented:
Nice work! I'll leave a minor comment in the diff.

axboe commented:
Oh, and please add signed-off-by line to your commit with your real name and email.

gsauthof commented:
Ok, I've added such a line.

axboe commented:
Can you make the other two changes I mentioned inline?

gsauthof commented:
> Can you make the other two changes I mentioned inline?

Hm, I didn't see those inline changes yesterday.
And I don't see them now.

Are you referring to github review comments?

Perhaps I'm having tomatoes on my eyes, but github doesn't show any for me, right now.
I hit F5 a few times and switched from the conversion to other tabs, but I don't see any review comments ...

axboe commented:
Maybe it's GH, sometimes it's a bit odd...

Attached as a screenshot :)

![image](https://user-images.githubusercontent.com/2619634/176890262-a652dbb8-abe1-446f-9d78-33602d14ceaa.jpeg)

axboe commented:
> You didn't click the "Submit Review" button again. That "Pending" state means only you who can see it.

Ah indeed - on the laptop now, and I see it in the corner. Don't even see it on the phone app. This is, honestly, the worst interface.

axboe commented:
Since we're getting into the weeds, I'd also do:

1) Use uint64_t rather than signed variables
2) Add fio_unlikely() annotation to rest to put it out-of-line, we don't expect to hit this

The latter removes a further 12 bytes for me on aarch64 (112 bytes -> 100 bytes) and does make a noticeable increase in performance.

axboe commented:
This would be easier to read as well:

```
void __fill_random_buf(void *buf, unsigned int len, uint64_t seed)
{
	uint64_t *b = buf;
	uint64_t *e = b + len / sizeof(*b);
	unsigned int rest = len % sizeof(*b);

	while (b != e) {
		*b = seed;
		seed *= GOLDEN_RATIO_PRIME;
		seed >>= 3;
		b++;
	}

	if (fio_unlikely(rest))
		__builtin_memcpy(e, &seed, rest);
}
```

axboe commented:
Some more data, since I got curious. With your change, the following synthetic test case:

```
./fio --name=test --ioengine=null --rw=write --bs=4k --refill_buffers=1 --size=100g
```

increases from about 1140K to 1165K here for my on my laptop. That's not a lot, perhaps it's better on x86-64. But it did lead me to experiment a bit. If we can just use crc32c where we have hw acceleration, then that might substantially improve the fill rate.

Did a quick hack to just use crc32c_arm64(), and then I get 2640K, or more than a doubling in fill rate. Now this obviously disregards the seed, which we will care about for verification. But for normal buffer refilling we really should not.

axboe commented:
https://git.kernel.dk/cgit/fio/commit/?h=random-fill&id=92568b01032c661ef54141e854ca920e4d18d156

I _think_ this should be fine here as we have no state at that point, and it gets me to ~2.7M rather than 1.14M IOPS for the above test case.

axboe commented:
Ran the same test case on an intel 12900K.

Current -git: 2090K
With patch in this PR: 2070K
Using crc32c as per referenced commit: 2670K

axboe commented:
Ignore the crc32c ramblings, pre-coffee nonsense. Won't work like described, obviously. And I don't think using the hw random is going to be useful. One thing that might be useful is having something that doesn't use integer multiplication and could rely on xor only, maybe some of the known hashing might be useful here. For now, let's just focus on the change at hand.

axboe commented:
Another idea - use GOLDEN_RATIO_64 and get rid of the shift. I _think_ that'd be good enough, and it should remove a dependency and increase the speed.

gsauthof commented on lib/rand.c:
```diff
@@ -97,29 +97,18 @@ void init_rand_seed(struct frand_state *state, uint64_t seed, bool use64)
-		seed *= GOLDEN_RATIO_PRIME;
-		seed >>= 3;
-	}
+    int64_t *b = buf;
+    int64_t *e = b  + len / sizeof *b;
+    unsigned int rest = len % sizeof *b;
+
+    for (int64_t *p = b; p != e; ++p) {
+        *p = seed;
+        seed *= GOLDEN_RATIO_PRIME;
+        seed >>= 3;
+    }
+
+    if (rest)
+        __builtin_memcpy(e, &seed, rest);
```
ok, I missed that tabs are being used in that project, I'll push an update

gsauthof commented on lib/rand.c:
```diff
@@ -97,29 +97,18 @@ void init_rand_seed(struct frand_state *state, uint64_t seed, bool use64)
+    int64_t *e = b  + len / sizeof *b;
+    unsigned int rest = len % sizeof *b;
+
+    for (int64_t *p = b; p != e; ++p) {
```
ok

gsauthof commented on lib/rand.c:
```diff
@@ -97,29 +97,18 @@ void init_rand_seed(struct frand_state *state, uint64_t seed, bool use64)
-		seed >>= 3;
-	}
+    int64_t *b = buf;
+    int64_t *e = b  + len / sizeof *b;
```
ok

gsauthof commented:
> Ran the same test case on an intel 12900K.

Yes, this is tricky to (micro-)benchmark.

In one profiling session I observed  the hot main loop shrinking from 10 to 6 instructions (and from 2 to 1 conditional instructions). However, when running on modern superscalar CPUs the CPU can - let's say - compensate for the additional instructions, to some degree.

As mentioned in my original post, the ~ 20 % fio throughput improvement I only saw on a kind of underpowered CPU.

---

I think I've addressed all the review comments.

I'll have to look into eliminating the shift by using GOLDEN_RATIO_64.
But perhaps this is something for another pull-request.

---

Another idea for further optimizations: use AES-NI instructions! (when available)

[Intel writes](https://www.intel.com/content/dam/doc/white-paper/advanced-encryption-standard-new-instructions-set-paper.pdf):

> The AES
instructions can also be used for all common uses of AES, including bulk encryption and
decryption using cipher modes such as CBC or CTR, data authentication using CBC-
MACs such as CMAC, _random number generation_ using algorithms such as CTR-DRBG,
and authenticated encryption using modes such as GCM.

(emphasis mine)

FWIW, a quick web search surfaced https://github.com/dragontamer/AESRand - I haven't tested it, but the README contains some food for thought.

axboe commented:
See the top two patches here, which are on top of your original:

https://git.kernel.dk/cgit/fio/log/?h=refill

which basically takes care of removing the dependency of the shift and multiplication, which nets me a nice win on actually modern CPUs. Randomness seems fine for what we need here.

Agree that using AES or similar would be the next step, that'd be a big improvement.

I'll pull yours on and rebase my two patches, at least that's a nice improvement over what we have.

axboe commented:
Janky test patch for aes random here: https://git.kernel.dk/cgit/fio/commit/?h=random-fill-aes&id=3208e9e30ac85363634177a8d9cb48ed6097dc44

On the 12900K we go from 2090K to 4230K, a bit more than double. Note that I haven't looked at the output at all, and the seeding etc is all wrong. Was mostly curious on how much faster we could go with AES, and I'm sure there's plenty of room for improvement still. It also assumes we're 16 byte aligned in size, etc.
